### DISCUSSION TITLE
Flask-socketio  Server communication problem

### DISCUSSION THREAD

**Discussion by Mohamad108** [miguelgrinberg/Flask-SocketIO#1609] [Q&A]
hi 
i 'm from Iran

i run my project on the system, it works fine, send and receive messages to the server
But
When I deploy within the site. The server response is incorrect, 
The server always sends the message socket.on ('connect')
Thanks for your help

![chat](https://user-images.githubusercontent.com/45655821/117576589-c0a12c00-b0fb-11eb-96d9-b65f0117d17c.png)
![chat-host](https://user-images.githubusercontent.com/45655821/117576598-c5fe7680-b0fb-11eb-853d-bad5f4f0a7d9.png)

# requirements:
              Flask==1.1.2
              Flask-SocketIO==5.0.1
              Flask-Session==0.3.2
              gunicorn==20.1.0
              eventlet==0.31.0

# app.py:
        from flask import Flask, render_template, request, redirect, url_for, session
        from flask_socketio import SocketIO, join_room, leave_room, emit
        #from flask_Sesstion import Session
        import eventlet
        eventlet.monkey_patch()
        
        app = Flask(__name__)
        app.debug = True
        app.config['SECRET_KEY'] = 'secret'
        app.config['SESSION_TYPE'] = 'filesystem'
        #Session(app)
        TempUser = []
        TempRoom = []
        
        socket = SocketIO(app, manage_session=False)
        
        @app.route('/', methods=['GET', 'POST'])
        def index():
            return render_template('index.html')
        
        @app.route('/chat', methods=['GET', 'POST'])
        def chat():
            if(request.method=='POST'):
                username = request.form['username']
                room = request.form['room']
                #Store the data in session
                session['username'] = username
                session['room'] = room
                return render_template('chat.html', session = session)
            else:
                if(session.get('username') is not None):
                    return render_template('chat.html', session = session)
                else:
                    return redirect(url_for('index'))
        
        @socket.on('join',namespace='/chat')
        def join(message):
            room = session.get('room')
            join_room(room)
            emit('status', {'msg': session.get('username') +' has entered the room.'}, room=room)
        
        @socket.on('text',namespace='/chat')
        def text(message):
            room = session.get('room')
            emit('message', {'msg': session.get('username') + ' : ' + message['msg']}, room=room)
        
        @socket.on('left',namespace='/chat')
        def left(message):
            room = session.get('room')
            username = session.get('username')
            leave_room(room)
            session.clear()
            emit('status', {'msg': username + ' has left the room.'}, room=room)
        
        if __name__ == '__main__':
            app.run()

# chat.html:
        <!DOCTYPE html>
        <html lang="en">
          <head>
            <title>Penguin Coders Chat</title>
            
            <link href="../static/bootstrap.min.css" rel="stylesheet" />
            <link href="../static/style.css" rel="stylesheet" />
            <script type="text/javascript" src="//code.jquery.com/jquery-3.2.1.slim.min.js"></script>
            <script type="text/javascript" src="//cdnjs.cloudflare.com/ajax/libs/socket.io/3.0.3/socket.io.min.js"></script>
            <script type="text/javascript" charset="utf-8">
                var socket;
                $(document).ready(function(){
                    socket = io.connect('http://'+document.domain + ':' + location.port +'/chat');
                    socket.on('connect', function() {
                        socket.emit('join', {});
                    });
                    socket.on('status', function(data) {
                        $('#chat').val($('#chat').val() + '<' + data.msg + '>\n');
                        $('#chat').scrollTop($('#chat')[0].scrollHeight);
                    });
                    socket.on('message', function(data) {
                        $('#chat').val($('#chat').val() + data.msg + '\n');
                        $('#chat').scrollTop($('#chat')[0].scrollHeight);
                    });
                    $('#send').click(function(e) {
                        text = $('#text').val();
                        $('#text').val('');
                        socket.emit('text',{'msg': text});
        
                    });
                });
                function leave_room() {
                    socket.emit('left',{},function() {
                        socket.disconnect();
                        window.location.href = "{{url_for('index')}}";
                    });
                }
            </script>
          </head>
          <body class="text-center">
            <div class="chatwindow">
              <h2 >Penguin Coders Chat</h2>
              <h2>Room : {{session['room']}}</h2><br/>
              <textarea id="chat" cols="70" rows="10" placeholder="No messages yet. Start one..."></textarea><br/><br/>
              <input type="text" id="text" size="60" placeholder="Enter your message here" />
              <button type="button" id="send" class="btn btn-success">SEND</button><br/><br/>
              <center><button type="button"  class="btn btn-danger" onclick="leave_room()"> Leave this Chat </button></center>
            </div>
          </body>
        </html>

**Comment by miguelgrinberg** [comment]
Please provide logs that show the problem. See the troubleshooting section of the documentation if you don't know how to enable the logs.

**Comment by Mohamad108** [comment]
# logs
raise RuntimeError('You need to use the eventlet server. '
RuntimeError: You need to use the eventlet server. See the Deployment section of the documentation for more information.

# Localhost:

    if __name__ == '__main__':
      import eventlet
      import eventlet.wsgi
      eventlet.wsgi.server(eventlet.listen(('', 8000)), app)

but 
How to use within the site??
How to set file passenger_wsgi??

**Comment by miguelgrinberg** [comment]
I don't know what you mean by passenger_wsgi. I assume this is a requirement of your deployment platform? You may need to ask them if they support eventlet.

**Comment by Mohamad108** [comment]
[ N 2021-05-10 20:24:20.8429 22254/T1 age/Cor/CoreMain.cpp:1117 ]: Checking whether to disconnect long-running connections for process 799, application /home/hironika/FlaskHironika (production)
[ N 2021-05-10 20:24:20.5406 22254/T1 age/Cor/CoreMain.cpp:1117 ]: Checking whether to disconnect long-running connections for process 799, application /home/hironika/FlaskHironika (production)
[ N 2021-05-10 17:24:18.4590 27247/T1 age/Cor/CoreMain.cpp:1117 ]: Checking whether to disconnect long-running connections for process 20640, application /home/hironika/FlaskHironika (production)
[ N 2021-05-10 17:24:18.2472 27247/T1 age/Cor/CoreMain.cpp:1117 ]: Checking whether to disconnect long-running connections for process 20640, application /home/hironika/FlaskHironika (production)

**Comment by miguelgrinberg** [comment]
I'm sorry, but I have no idea what you are showing me.

**Comment by bluthen** [comment]
I think those are from Apache. Perhaps apache wsgi set to have several workers, and it is checking if flask is still alive for each worker

**Comment by Mohamad108** [comment]
Not a problem of the program
This problem is solved when I run the program with socketio.run (app)
But when I transfer it to the host, it is executed by passenger_wsgi, which is dedicated to the host
I think the problem is with the setting of passenger_wsgi files

thank you very much

**Comment by miguelgrinberg** [comment]
What web server does this passenger_wsgi thing uses? You can't just use any web server, there are some requirements that many web servers (say Apache) do not have.
